src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs

```diff
@@ -38,9 +40,11 @@ public ConnectionInitializer(string applicationName)
         public ConnectionDescription InitializeConnection(IConnection connection, CancellationToken cancellationToken)
         {
             Ensure.IsNotNull(connection, nameof(connection));
-
-            var isMasterProtocol = CreateIsMasterProtocol();
-            var isMasterResult = new IsMasterResult(isMasterProtocol.Execute(connection, cancellationToken));
+            var isMasterCommand = AddAuthenticatorCustomizationsToIsMasterCommand(
+                AddClientDocumentToIsMasterCommand(CreateIsMasterCommand(), _clientDocument),
+                connection.Settings.Authenticators);
```

Comment:
I find these three lines confusing. How about:

    var isMasterCommand = CreateIsMasterCommand();

and let CreateIsMasterCommand handle the details?

A CreateIsMasterCommand helper method would also reduce duplication in InitializeConnectionAsync.

You could even call CreateIsMasterCommand from CreateIsMasterProtocol to reduce duplication even further.

Reply:
I do agree it is a bit nested  [As an aside, a pity that C# doesn't have a thread-first macro to help readability in such situations (https://clojure.org/guides/threading_macros)--then we could do something like 
```
-> CreateIsMasterCommand() 
      AddClientDocumentToIsMasterCommand(_clientDocument)
      AddAuthenticatorCustomizationsToIsMasterCommand(connection.settings.Authenticators);
```
]

Unfortunately, currently there's a bit of naming collision as  `DefaultAuthenticator` also relies on the static  `CreateIsMasterCommand()` to create no-frills "isMaster" command. This is partially because my initial intent was to isolate the responsibilities of each function as much as possible (and allow reuse when possible).
How do you feel about about bundling the 3 lines into a `CreateInitialIsMasterCommand()`?
Alternatively, the IsMaster functionality could also be factored into into a separate helper class.
If we do this, the three lines indicated could be placed into a `CreateIsMasterCommand` as suggested since we'd no longer have a naming collision while still sharing functionality with the `DefaultAuthenticator` class.

Reply:
Added IsMasterHelper class as discussed. Created a new method for the 3 lines as discussed.